src/dill/gauges/GaugeV2.sol

```diff
@@ -0,0 +1,715 @@
+                    thisStake.ending_timestamp;
+
+                // Get the weighted-average lock_multiplier
+                uint256 numerator = (lock_multiplier * timeBeforeExpiry) +
+                    (_MultiplierPrecision * timeAfterExpiry);
+                lock_multiplier =
+                    numerator /
+                    (timeBeforeExpiry + timeAfterExpiry);
+            }
+            // Otherwise, it needs to just be 1x
+            else {
+                lock_multiplier = _MultiplierPrecision;
+            }
+        } else {
+            uint256 elapsedSeconds = block.timestamp - lastRewardClaimTime;
+            if (elapsedSeconds > 0) {
+                lock_multiplier = thisStake.isPermanentlyLocked
+                    ? lockMaxMultiplier
+                    : _averageDecayedLockMultiplier(account, elapsedSeconds);
+                _lastUsedMultiplier[account] =
+                    _lastUsedMultiplier[account] -
+                    (elapsedSeconds - 1) *
+                    multiplierDecayPerSecond;
+            }
+        }
+        uint256 liquidity = thisStake.liquidity;
+        uint256 combined_boosted_amount = (liquidity * lock_multiplier) /
+            _MultiplierPrecision;
+        lockBoostedDerivedBal = lockBoostedDerivedBal + combined_boosted_amount;
+
+        return dillBoostedDerivedBal + lockBoostedDerivedBal;
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function kick(address account) public {
+        uint256 _derivedBalance = derivedBalances[account];
+        derivedSupply = derivedSupply - _derivedBalance;
+        _derivedBalance = derivedBalance(account);
+        derivedBalances[account] = _derivedBalance;
+        derivedSupply = derivedSupply + _derivedBalance;
+    }
+
+    // Simple Deposits
+
+    function depositAll() external {
+        require(TOKEN.balanceOf(msg.sender) > 0, "Cannot deposit 0");
+        _deposit(TOKEN.balanceOf(msg.sender), msg.sender, 0, 0, false, true);
+    }
+
+    function depositFor(uint256 amount, address account) external {
+        require(amount > 0, "Cannot deposit 0");
+        require(
+            stakingDelegates[account][msg.sender],
+            "Only registerd delegates can deposit for their deligator"
+        );
+        _deposit(amount, account, 0, 0, false, true);
+    }
+
+    function deposit(uint256 amount) external {
+        require(amount > 0, "Cannot deposit 0");
+        _deposit(amount, msg.sender, 0, 0, false, true);
+    }
+
+    function depositAllAndLock(uint256 secs, bool isPermanentlyLocked)
+        external
+        lockable(secs)
+    {
+        require(TOKEN.balanceOf(msg.sender) > 0, "Cannot stake 0");
+        LockedStake memory thisStake = _lockedStakes[msg.sender];
+        // Check if stake already exists and if it is unlocked
+        if (thisStake.liquidity > 0) {
+            require(
+                (!stakesUnlocked || !stakesUnlockedForAccount[msg.sender]) &&
```

Comment:
I believe it should be an OR for the last evaulation. Any of these conditions being satisfied means the user can withdraw their stake and should do so before locking again.

Also, per my feedback on our initial review, I don't see functionality which allows users to escape the boost decay (by extendin an existing lock for example).